ding/envs/env_manager/env_supervisor.py

```diff
@@ -0,0 +1,552 @@
+                callback recursively.
+        """
+        self._set_shared_obs(payload=payload)
+        self.change_state(payload=payload)
+        if payload.method == "reset":
+            return self._recv_reset_callback(payload=payload, remain_payloads=remain_payloads)
+        elif payload.method == "step":
+            return self._recv_step_callback(payload=payload, remain_payloads=remain_payloads)
+        return payload
+
+    def _set_shared_obs(self, payload: RecvPayload):
+        if self._obs_buffers is None:
+            return
+        if payload.method == "reset" and payload.err is None:
+            payload.data = self._obs_buffers[payload.proc_id].get()
+        elif payload.method == "step" and payload.err is None:
+            payload.data._replace(obs=self._obs_buffers[payload.proc_id].get())
+
+    def _recv_reset_callback(
+            self, payload: RecvPayload, remain_payloads: Optional[Dict[str, SendPayload]] = None
+    ) -> RecvPayload:
+        assert payload.method == "reset", "Recv error callback({}) in reset callback!".format(payload.method)
+        if remain_payloads is None:
+            remain_payloads = {}
+        env_id = payload.proc_id
+        if payload.err:
+            self._retry_times[env_id] += 1
+            if self._retry_times[env_id] > self._max_try - 1:
+                self.shutdown(5)
+                raise RuntimeError(
+                    "Env {} reset has exceeded max_try({}), and the latest exception is: {}".format(
+                        env_id, self._max_try, payload.err
+                    )
+                )
+            if self._retry_waiting_time:
+                sleep(self._retry_waiting_time)
+            if self._retry_type == EnvRetryType.RENEW:
+                self._children[env_id].restart()
+            send_payloads = self._reset(env_id)
+            for p in send_payloads:
+                remain_payloads[p.req_id] = p
+        else:
+            self._retry_times[env_id] = 0
+            self._ready_obs[env_id] = payload.data
+        return payload
+
+    def _recv_step_callback(
+            self, payload: RecvPayload, remain_payloads: Optional[Dict[str, SendPayload]] = None
+    ) -> RecvPayload:
+        assert payload.method == "step", "Recv error callback({}) in step callback!".format(payload.method)
+        if remain_payloads is None:
+            remain_payloads = {}
+        if payload.err:
+            send_payloads = self._reset(payload.proc_id)
+            for p in send_payloads:
+                remain_payloads[p.req_id] = p
+            info = {"abnormal": True, "err": payload.err}
+            payload.data = tnp.array(
+                {
+                    'obs': None,
+                    'reward': None,
+                    'done': None,
+                    'info': info,
+                    'env_id': payload.proc_id
+                }
+            )
+        else:
+            obs, reward, done, info, *_ = payload.data
```

Comment:
why `*_` here

Reply:
I found that some environments still have variables after info, for example smac